cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcClient.java

```diff
@@ -70,13 +88,83 @@
   private final DataProcConf conf;
   private final ClusterControllerClient client;
   private final Compute compute;
+  private final String projectId;
 
   public static DataProcClient fromConf(DataProcConf conf) throws IOException, GeneralSecurityException {
-    ClusterControllerClient client = ClusterControllerClient.create(conf.getControllerSettings());
-    return new DataProcClient(conf, client, conf.getCompute());
+    String projectId = conf.getProjectId();
+    if (projectId == null) {
+      projectId = getProjectId();
+    }
+
+    String accountKey = conf.getAccountKey();
+    ClusterControllerClient client = getClusterControllerClient(accountKey);
+    Compute compute = getCompute(accountKey);
+
+    return new DataProcClient(projectId, conf, client, compute);
+  }
+
+  private static String getProjectId() throws IOException {
+    URL url = new URL("http://metadata.google.internal/computeMetadata/v1/project/project-id");
+    HttpURLConnection connection = null;
+    try {
+      connection = (HttpURLConnection) url.openConnection();
+      connection.setRequestProperty("Metadata-Flavor", "Google");
+      connection.connect();
+      try (Reader reader = new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8)) {
+        return CharStreams.toString(reader);
+      }
+    } catch (IOException e) {
+      throw new IOException("Unable to get project id from the environment. "
+                              + "Please explicitly set the project id and account key.", e);
+    } finally {
+      if (connection != null) {
+        connection.disconnect();
+      }
+    }
+  }
+
+  private static ClusterControllerClient getClusterControllerClient(@Nullable String accountKey) throws IOException {
+    GoogleCredentials credentials;
+    if (accountKey != null) {
+      try (InputStream is = new ByteArrayInputStream(accountKey.getBytes(StandardCharsets.UTF_8))) {
+        credentials = GoogleCredentials.fromStream(is);
+      }
+    } else {
+      try {
+        credentials = ComputeEngineCredentials.create();
+        credentials.refreshAccessToken();
+      } catch (IOException e) {
+        throw new IOException("Unable to get credentials from the environment. "
+                                + "Please explicitly set the account key.", e);
+      }
+    }
+    CredentialsProvider credentialsProvider = FixedCredentialsProvider.create(credentials);
+
+    ClusterControllerSettings controllerSettings = ClusterControllerSettings.newBuilder()
+      .setCredentialsProvider(credentialsProvider)
+      .build();
+    return ClusterControllerClient.create(controllerSettings);
+  }
+
+  private static Compute getCompute(@Nullable String accountKey) throws GeneralSecurityException, IOException {
+    HttpTransport httpTransport = GoogleNetHttpTransport.newTrustedTransport();
+
+    GoogleCredential credential;
```

Comment:
just curious why we have to read the account key twice, once to construct `Credentials` and now to create `Credential`,  isn't there a way to convert `Credentials` to `Credential` ?

Reply:
It would be nice if there were, but I didn't see a way. The dataproc and compute engine APIs seem to use different methods of getting credentials.